src/main/scala/combinator/BanditSelector.scala

```diff
@@ -0,0 +1,449 @@
+
+  // number of neighborhoods marked as tabu
+  protected var nTabu = 0
+
+  // false if a neighborhood is marked as tabu
+  protected val authorizedNeighborhood: Array[Boolean] = Array.fill(neighborhoods.length)(true)
+
+  // weight associated to each neighborhood
+  protected val weights: Array[Double] = Array.fill(neighborhoods.length)(1.0)
+
+  // sum of weights of neighborhoods that are not marked as tabu
+  protected var sumWeightsValidNeighborhoods: Double = neighborhoods.length
+
+  // for each neighborhood, its index in the list
+  protected var neighborhoodIdx: mutable.HashMap[Neighborhood, Int] = mutable.HashMap.empty
+
+  // list of stats for each neighborhood. One stat is collected per move
+  protected var stats: Array[ListBuffer[NeighborhoodStats]] =
+    Array.fill(neighborhoods.length)(ListBuffer())
+  protected val rand = new Random(seed)
+
+  // number of time each neighborhood was selected
+  protected val nSelected: Array[Int] = Array.fill(neighborhoods.length)(0)
+
+  // index of the last selected neighborhood
+  protected var lastSelectedIdx: Int = -1
+
+  // Populate the map
+  for (i <- neighborhoods.indices) {
+    neighborhoodIdx += (neighborhoods(i) -> i)
+  }
+
+  protected var maxSlope             = 1.0 // stores (and updates) the maximum slope ever observed
+  protected var maxRunTimeNano: Long = 1   // max run time experienced by a neighborhood
+
+  // TODO next steps to slightly speeds the selection:
+  //  1. use sparse-set to maintain the neighborhoods that are not marked as tabu
+  //  2. use sparse-set to maintain the neighborhoods that have been selected at least once since the last weight update (only them must have their weight updated)
+
+  /** The method that provides a neighborhood.
+    *
+    * @return
+    *   Some(n) if a neighborhood is available or None if the neighborhoods are exhausted
+    */
+  def getNextNeighborhood: Option[Neighborhood]
+
+  /** The method that computes a reward associated to a neighborhood.
+    *
+    * @return
+    *   Some(n) if a neighborhood is available or None if the neighborhoods are exhausted
+    */
+  def reward(runStat: NeighborhoodStats, neighborhood: Neighborhood): Double
+
+  /** Resets the selector. This resets the tabu list and update the weights if the learning scheme
+    * is set to
+    *   - after every descent.
+    *   - after n moves and the current.
+    */
+  override def reset(): Unit = {
+    resetTabu()
+    learningScheme match {
+      case AfterEveryDescent =>
+        updateWeights()
+        clearStats()
+      case afterNMoves @ AfterNMoves(_) =>
+        if (afterNMoves.nMoves > 0) {
+          updateWeights()
+          afterNMoves.resetCounter()
+          clearStats()
+        }
+      case AfterEveryMove =>
+    }
+  }
+
+  /** Clears the stats associated to all neighborhoods.
+    */
+  private def clearStats(): Unit = {
+    for (idx <- stats.indices) {
+      stats(idx).clear()
```

Comment:
```suggestion
   stats.foreach(_.clear())
```
i don't think that it changes much though